src/preprocessing/shapes/triangle_mesh.jl

```diff
@@ -0,0 +1,138 @@
+# This is the data format returned by `load(file)` when used with `.stl` files
+struct TriangleMesh{NDIMS, ELTYPE} <: Shapes{NDIMS}
+    vertices          :: Vector{SVector{NDIMS, ELTYPE}}
+    face_vertices     :: Vector{Tuple{SVector{NDIMS, ELTYPE}, SVector{NDIMS, ELTYPE}, SVector{NDIMS, ELTYPE}}}
+    face_vertices_ids :: Vector{NTuple{3, Int}}
+    face_edges_ids    :: Vector{NTuple{3, Int}}
+    edge_vertices_ids :: Vector{NTuple{2, Int}}
+    normals_vertex    :: Vector{SVector{NDIMS, ELTYPE}}
+    normals_edge      :: Vector{SVector{NDIMS, ELTYPE}}
+    normals_face      :: Vector{SVector{NDIMS, ELTYPE}}
+    min_corner        :: SVector{NDIMS, ELTYPE}
+    max_corner        :: SVector{NDIMS, ELTYPE}
+
+    function TriangleMesh(face_vertices, normals_face, vertices)
+        NDIMS = 3
+        ELTYPE = eltype(first(normals_face))
+        n_faces = length(normals_face)
+
+        face_vertices_ids = fill((0, 0, 0), n_faces)
+
+        @threaded face_vertices for i in 1:n_faces
+            v1 = face_vertices[i][1]
+            v2 = face_vertices[i][2]
+            v3 = face_vertices[i][3]
+
+            # TODO: This part is about 90% of the runtime.
+            # Vertex IDs are only needed for the hierarchical winding.
+            vertex_id1 = findfirst(x -> v1 == x, vertices)
+            vertex_id2 = findfirst(x -> v2 == x, vertices)
+            vertex_id3 = findfirst(x -> v3 == x, vertices)
+
+            face_vertices_ids[i] = (vertex_id1, vertex_id2, vertex_id3)
+        end
```

Comment:
You build the `vertices` array yourself.
```julia
        face_vertices[i + 1] = (v1, v2, v3)

        vertices[3 * i + 1] = v1
        vertices[3 * i + 2] = v2
        vertices[3 * i + 3] = v3
```
And then you have to find the indices again? This seems very unnecessary to me.
Why don't you make `face_vertices` an `NTuple{3, Int}`, which points to the vertex IDs?
Then you can also get rid of duplicate vertices. I guess each vertex appears once for each adjacent triangle.

Reply:
>  I guess each vertex appears once for each adjacent triangle.

No, that is the problem. The stl-file only stores the coordinates of the triangle vertices. That is, we do not know which triangles are adjacent or which triangles do share the same vertex. That means, each vertex ID is unique but there are vertices with the same coordinates.

Reply:
No that's what I meant. You each vertex appears again for each adjacent triangle. But in the code I quoted above, you can check if this vertex already exists, and skip it in this case. But if you don't care if you have duplicate vertices, then don't do this.

Reply:
My idea doesn't work, as discussed, because the vertices need to be unique.